src/MongoDB.Driver.Core/Core/Configuration/ClusterSettings.cs

```diff
@@ -252,5 +318,39 @@ public IServerSelector PostServerSelector
                 schemaMap: Optional.Create(schemaMap.WithDefault(_schemaMap)),
                 scheme: scheme.WithDefault(_scheme));
         }
+
+        // internal methods
+        internal ClusterType GetClusterType()
+        {
+#pragma warning disable 618
+            if (_connectionModeSwitch == ConnectionModeSwitch.UseDirectConnection)
+#pragma warning restore 618
+            {
+                if (!_directConnection.GetValueOrDefault() && _replicaSetName != null)
```

Comment:
Can we reverse the order of the tests?

    if (_replicaSetName != null && !_directConnection.GetValueOrDefault())

In English it's like saying "assume cluster type is ReplicaSet when _replicaSetName is set EXCEPT for direct connections".